crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -159,6 +159,53 @@ impl From<IndexeddbCryptoStoreError> for CryptoStoreError {
 
 type Result<A, E = IndexeddbCryptoStoreError> = std::result::Result<A, E>;
 
+/// Defines an operation to perform on the database
+enum DbOperation {
+    PutKeyVal { key: JsValue, value: JsValue },
+    Delete(JsValue),
+}
+/// A structs that represents all the operations that need to be done to the
+/// database when save_changes is called.
+/// The idea is to do all the serialization and encryption before the
+/// transaction, and then just do the actual Indexeddb operations in the
+/// transaction.
+struct IndexeddbChangesKeyValue {
+    /// A map of the object store names to the operations to perform on that
+    /// store.
+    store_to_key_values: HashMap<&'static str, Vec<DbOperation>>,
+}
+
+impl IndexeddbChangesKeyValue {
+    fn new() -> Self {
+        Self { store_to_key_values: HashMap::new() }
+    }
+
+    fn get(&mut self, store: &'static str) -> &mut Vec<DbOperation> {
```

Comment:
There's a possibly better API where calling `get()` returns something that would hold and manipulate the `&mut Vec<DbOperation>`, and would avoid lots of boilerplate at call site:

```rust
struct PendingStoreChanges<'a> {
  operations: &'a mut Vec<DbOperation>
}

impl <'a> PendingChanges<'a> {
  fn put(&mut self, key: JsValue, value: JsValue) {
    self.operations.push(DbOperation::Put { key, value });
  }
}

impl IndexeddbChangesKeyValue {
  fn get(&mut self, store: &'static str) -> PendingStoreChanges<'_> {
    PendingChanges { operations: self.store_to_key_values.entry(store).or_default() }
  }
}

// call sites
let core = indexeddb_changes_key_value.get(CORE);
core.put(jsKey, jsValue);
// etc
```

So if you plan to reuse this scheme somewhere else, I would recommend doing this change :)

(Also, the `put` method could even take an `impl Serialize` argument and embed the `self.serialize_value(...)` instead of requiring the caller to do it)